### PR TITLE
Fix MKL Conv1D crash with dynamic batch size

### PR DESCRIPTION
**Pull request by chinanuj**
This PR fixes Issue #105331 where tf.nn.conv1d_transpose crashes when using the MKL backend with a dynamic batch size (-1). Root Cause : The function MakeInputTfShape was using tensor::MakeShape directly on the input tensor. tensor::MakeShape throws an INVALID_ARGUMENT error if it encounters a -1 dimension. Fix : Added logic to detect if the batch dimension is -1. If detected, the code now infers the correct batch size from the diff_dst (gradient) tensor before creating the shape. Note : I do not have an Intel MKL environment to verify this runtime fix locally, but the logic follows the standard pattern for handling dynamic shapes in other MKL kernels.

### CODE REVIEW

**Comment by chinanuj** [comment]
@google-cla check

**Review by mihaimaruseac — DISMISSED** [review]

**Comment by chinanuj** [comment]
Hello @mihaimaruseac , Thanks for the earlier review.

I have pushed a fix for the Linux x86 CPU failure. The previous crash was caused by a data type mismatch (int32 vs int64) inside MakeInputTfShape.

I updated the logic to strictly check dtype() and handle both DT_INT32 and DT_INT64 inputs before processing the dynamic batch size. This should resolve the regression.

Could you please take a look and run on the tests? Thanks!

File: tensorflow/core/kernels/mkl/mkl_conv_grad_input_ops.cc [hunk 1]
```diff
@@ -567,14 +567,54 @@ class MklConvCustomBackpropInputOp
         << "ConvBackpropInput: input should not be in MKL Layout";
   }
 
-  // Get TensorFlow shape of input tensor.
+// Get TensorFlow shape of input tensor.
```
Comment (mihaimaruseac) [inline tensorflow/core/kernels/mkl/mkl_conv_grad_input_ops.cc]: This should maintain the original indentation

File: tensorflow/core/kernels/mkl/mkl_conv_grad_input_ops.cc [hunk 2]
```diff
@@ -567,14 +567,54 @@ class MklConvCustomBackpropInputOp
...
   TensorShape MakeInputTfShape(OpKernelContext* context,
                                const Tensor& input_tensor) {
     TensorShape input_tf_shape;
     CHECK_EQ(TensorShapeUtils::IsVector(input_tensor.shape()), true);
-    // Conv[2D|3D]BackpropInputV2 supports both DT_INT32 and DT_INT64
-    // output_shape tensor::MakeShape is able to handle both DT_INT32 and
-    // DT_INT64 for input_tensor.
+
+    // FIX START: Handle dynamic batch size (-1) supporting both int32 and int64
```
Comment (mihaimaruseac) [inline tensorflow/core/kernels/mkl/mkl_conv_grad_input_ops.cc]: These comment don't help. Please remove AI generated comments if they are not helpful

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

**Comment by chinanuj** [comment]
Hello @mihaimaruseac , I've fixed the indentation to match the style guide and dropped the verbose comments. Please take a look if it works fine .

File: tensorflow/core/kernels/mkl/mkl_conv_grad_input_ops.cc [hunk 3]
```diff
@@ -572,9 +572,43 @@ class MklConvCustomBackpropInputOp
...
+      };
+
+      if (input_tensor.dtype() == DT_INT32) {
+        auto shape_vec = input_tensor.flat<int32>();
+        if (shape_vec(0) == -1) {
+          is_dynamic = true;
+          int64 batch_size = get_batch_size();
+          for (int i = 0; i < input_tensor.NumElements(); i++) {
+            int32 val = shape_vec(i);
+            explicit_shape.push_back(val == -1 ? batch_size : val);
```
Comment (mihaimaruseac) [inline tensorflow/core/kernels/mkl/mkl_conv_grad_input_ops.cc]: I would extract this repeated code to a template function and just call it from here

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

**Comment by chinanuj** [comment]
@mihaimaruseac Agreed, definitely that's a better way. I have fixed the code now and added a template to reduce code repeatability. Please take a look!

File: tensorflow/core/kernels/mkl/mkl_conv_grad_input_ops.cc [hunk 4]
```diff
@@ -567,14 +568,44 @@ class MklConvCustomBackpropInputOp
         << "ConvBackpropInput: input should not be in MKL Layout";
   }
 
+  // Handle dynamic shape of input tensor.
+  template <typename T>
+  bool HandleDynamicShape(const Tensor& input_tensor, OpKernelContext* context,
```
Comment (mihaimaruseac) [inline tensorflow/core/kernels/mkl/mkl_conv_grad_input_ops.cc]: Why not make this return a vector directly and return an empty one if you don't handle it?

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

**Comment by chinanuj** [comment]
@mihaimaruseac Got it . I have updated the Template , Hope it works fine now . Thanks for the guidance!

**Review by mihaimaruseac — DISMISSED** [review]

**Comment by chinanuj** [comment]
@mihaimaruseac Apologies for the extra noise—it turns out the compiler needed int64 instead of int64_t , due to this there was a Build fail . I've pushed the fix. Please take a look!

**Review by mihaimaruseac — DISMISSED** [review]

**Review by mihaimaruseac — APPROVED** [review]

**Comment by chinanuj** [comment]
@mihaimaruseac It looks like all the checks have finally passed! 
However, the internal feedback/copybara check failed (infrastructure error). Could you please retry the sync so it can merge? Thanks!
